RecoPixelVertexing/PixelTrackFitting/interface/RiemannFit.h

```diff
@@ -0,0 +1,947 @@
+    const MatrixNd H = MatrixXd::Identity(n, n).rowwise() - weight.transpose();
+    const MatrixNx3d s_v = H * p3D.transpose();
+
+    MatrixNd D_[3][3];  // cov(s_v)
+    {
+      D_[0][0] = (H * C[0][0] * H.transpose()).cwiseProduct(W);
+      D_[0][1] = (H * C[0][1] * H.transpose()).cwiseProduct(W);
+      D_[0][2] = (H * C[0][2] * H.transpose()).cwiseProduct(W);
+      D_[1][1] = (H * C[1][1] * H.transpose()).cwiseProduct(W);
+      D_[1][2] = (H * C[1][2] * H.transpose()).cwiseProduct(W);
+      D_[2][2] = (H * C[2][2] * H.transpose()).cwiseProduct(W);
+      D_[1][0] = D_[0][1].transpose();
+      D_[2][0] = D_[0][2].transpose();
+      D_[2][1] = D_[1][2].transpose();
+    }
+
+    constexpr u_int nu[6][2] = {{0, 0}, {0, 1}, {0, 2}, {1, 1}, {1, 2}, {2, 2}};
+
+    Matrix6d E;  // cov matrix of the 6 independent elements of A
+    for (u_int a = 0; a < 6; ++a) {
+      const u_int i = nu[a][0], j = nu[a][1];
+      for (u_int b = a; b < 6; ++b) {
+        const u_int k = nu[b][0], l = nu[b][1];
+        VectorNd t0(n);
+        VectorNd t1(n);
+        if (l == k) {
+          t0 = 2. * D_[j][l] * s_v.col(l);
+          if (i == j)
+            t1 = t0;
+          else
+            t1 = 2. * D_[i][l] * s_v.col(l);
+        } else {
+          t0 = D_[j][l] * s_v.col(k) + D_[j][k] * s_v.col(l);
+          if (i == j)
+            t1 = t0;
+          else
+            t1 = D_[i][l] * s_v.col(k) + D_[i][k] * s_v.col(l);
+        }
+
+        if (i == j)
+          E(a, b) = 0. + s_v.col(i).transpose() * (t0 + t1);
+        else
+          E(a, b) = 0. + (s_v.col(i).transpose() * t0) + (s_v.col(j).transpose() * t1);
+        if (b != a) E(b, a) = E(a, b);
+      }
+    }
+
+    Matrix<double, 3, 6> J2;  // Jacobian of min_eigen() (numerically computed)
+    for (u_int a = 0; a < 6; ++a) {
+      const u_int i = nu[a][0], j = nu[a][1];
+      Matrix3d Delta = Matrix3d::Zero();
+      Delta(i, j) = Delta(j, i) = abs(A(i, j) * d);
+      J2.col(a) = min_eigen3D_fast(A + Delta);
+      const int sign = (J2.col(a)(2) > 0) ? 1 : -1;
+      J2.col(a) = (J2.col(a) * sign - v) / Delta(i, j);
+    }
+
+    Matrix4d Cvc;  // joint cov matrix of (v0,v1,v2,c)
+    {
+      Matrix3d t0 = J2 * E * J2.transpose();
+      Vector3d t1 = -t0 * r0;
+      Cvc.block(0, 0, 3, 3) = t0;
+      Cvc.block(0, 3, 3, 1) = t1;
+      Cvc.block(3, 0, 1, 3) = t1.transpose();
+      Cvc(3, 3) =
+          (v.transpose() * C0 * v) + (C0.cwiseProduct(t0)).sum() + (r0.transpose() * t0 * r0);
+    }
+
+    Matrix<double, 3, 4> J3;  // Jacobian (v0,v1,v2,c)->(X0,Y0,R)
+    {
+      const double t = 1. / h;
+      J3 << -v2x2_inv, 0, v(0) * sqr(v2x2_inv) * 2., 0, 0, -v2x2_inv, v(1) * sqr(v2x2_inv) * 2., 0,
+          0, 0, -h * sqr(v2x2_inv) * 2. - (2. * c + v(2)) * v2x2_inv * t, -t;
+    }
+
+    const RowVector2Nd Jq = mc.transpose() * s * 1. / n;  // var(q)
+
+    Matrix3d cov_uvr = J3 * Cvc * J3.transpose() * sqr(s_inv)  // cov(X0,Y0,R)
```

Comment:
do you know which Eigen is right ?

Reply:
@fwyzard Good question! I investigated a bit and now think that it's a bug in eigen 3.2.9.

`sqr(s_inv)` is a scalar, 
`J3 * Cvc * J3.transpose()` is a 3x3 matrix
`(par_uvr_ * par_uvr_.transpose())` is a 3x3 matrix
`Jq * V * Jq.transpose()` is a 1x1 matrix

and  eigen 3.2.9 doesn't seem to recognise that the multiplication by the 1x1 matrix is the same as scalar multiplication

It compiles just adding the line `double jqProd = Jq * V * Jq.transpose();`  and using this double in the calculation of `cov_uvr`. Since it doesn't cost much and makes the code compatible with the eigen version in the latest LCG release, it might be worthwhile